cmd/gonomics/usage.go

```diff
@@ -0,0 +1,244 @@
+//# GONOMICS: testse
+
+// Package main does stuff
+package main
+
+import (
+	"fmt"
+	"github.com/vertgenlab/gonomics/fileio"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"runtime"
+	"sort"
+	"strings"
+	"text/tabwriter"
+)
+
+var Reset = "\033[0m"
+var Red = "\033[31m"
+var Green = "\033[32m"
+var Yellow = "\033[33m"
+var Blue = "\033[34m"
+var Purple = "\033[35m"
+var Cyan = "\033[36m"
+var Gray = "\033[37m"
+var White = "\033[97m"
+
+const hline = "━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━"
+
+// init prevents color characters if running Windows.
+// Windows does not allow colored terminal output by
+// default so it just adds a bunch of characters to
+// the print statements.
+func init() {
+
+	if runtime.GOOS == "windows" {
+		Reset = ""
+		Red = ""
+		Green = ""
+		Yellow = ""
+		Blue = ""
+		Purple = ""
+		Cyan = ""
+		Gray = ""
+		White = ""
+	}
+}
+
+type CmdInfo struct {
+	Name  string
+	Usage string
+	Group string
+}
+
+func usage() {
+	cache := getCache()
+
+	fmt.Print(
+		Yellow + "gonomics - A collection of tools that use the gonomics core library.\n\n" +
+			"Usage: gonomics <command> [options]\n\n" + Reset)
+	fmt.Println(Red + "Commands:" + Reset)
+	fmt.Print(hline)
+	printCmdList(cache)
+	fmt.Println(hline)
+}
+
+// getCache returns a cache file ('go bin path'/.cmdcache) listing the cmd usage statements.
+// Retrieves the file if it exists, else builds a new cache.
+func getCache() (cache *os.File) {
+	binPath, _ := getBin()
+
+	cmdStat, _ := os.Stat(os.Args[0])                          // stat binary for this file
+	cacheStat, cacheStatErr := os.Stat(binPath + "/.cmdcache") // stat cache file
+
+	// Build a cache if cache does not exist OR if the cache is older than the gonomics cmd binary
+	if os.IsNotExist(cacheStatErr) || cmdStat.ModTime().After(cacheStat.ModTime()) {
+		buildCmdCache(binPath)
+	}
+	file, err := os.Open(binPath + "/.cmdcache")
+	if err != nil {
+		log.Panic(err)
+	}
+	return file
+}
+
+// printCmdList prints the cache file
+func printCmdList(cache *os.File) {
+	toPrint, err := ioutil.ReadAll(cache)
+	cache.Close()
+	if err != nil {
+		log.Panic(err)
+	}
+	fmt.Print(string(toPrint))
+}
+
+// buildCmdCache creates a cache file listing the cmd usage statements.
+// TODO find a way to store cache in gonomics binary file
+func buildCmdCache(binPath string) {
+	fmt.Println("...Building Cache...")
+	cmdMap := getGonomicsCmds()
+
+	cmds := make([]string, 0, len(cmdMap))
+	for key := range cmdMap {
+		cmds = append(cmds, key)
+	}
+	sort.Slice(cmds, func(i, j int) bool { return cmds[i] < cmds[j] })
+
+	cmdSourcePath := os.Getenv("GOPATH") + "/src/github.com/vertgenlab/gonomics/cmd/"
```

Comment:
Yes. I am not sure how to get around it. In previous versions of go, you needed to have your source code in GOPATH, but now you can put your source code anywhere you want. I have not been able to think of a good way to figure out where the source code is (it may not even exist if we end up a binary version of gonomics). Ideally this would run whenever you run `go install ./...` so we would know exactly where the source code is AND when the cache needs to be updated, but I have not found a way to do this (or know if it is possible). Open to any suggestions.

Reply:
Though, I did add a check now to default to $HOME/go if GOPATH is not set.